staging/src/k8s.io/client-go/tools/events/event_broadcaster.go

```diff
@@ -0,0 +1,290 @@
+			time.Sleep(sleepDuration)
+		}
+	}
+
+}
+
+func recordEvent(sink EventSink, event *v1beta1.Event, isEventSerie bool) bool {
+	var newEvent *v1beta1.Event
+	var err error
+	if isEventSerie {
+
+		patch, err := createPatchBytes(event)
+		if err != nil {
+			klog.Errorf("Unable to calculate diff, no merge is possible: %v", err)
+			return false
+		}
+		newEvent, err = sink.Patch(event, patch)
+	}
+	// Update can fail because the event may have been removed and it no longer exists.
+	if !isEventSerie || (isEventSerie && util.IsKeyNotFoundError(err)) {
+		// Making sure that ResourceVersion is empty on creation
+		event.ResourceVersion = ""
+		newEvent, err = sink.Create(event)
+	}
+	if err == nil {
+
+		event = newEvent
+		return true
+	}
+	// If we can't contact the server, then hold everything while we keep trying.
+	// Otherwise, something about the event is malformed and we should abandon it.
+	switch err.(type) {
+	case *restclient.RequestConstructionError:
+		// We will construct the request the same next time, so don't keep trying.
+		klog.Errorf("Unable to construct event '%#v': '%v' (will not retry!)", event, err)
+		return true
+	case *errors.StatusError:
+		if errors.IsAlreadyExists(err) {
+			klog.V(5).Infof("Server rejected event '%#v': '%v' (will not retry!)", event, err)
+		} else {
+			klog.Errorf("Server rejected event '%#v': '%v' (will not retry!)", event, err)
+		}
+		return true
+	case *errors.UnexpectedObjectError:
+		// We don't expect this; it implies the server's response didn't match a
+		// known pattern. Go ahead and retry.
+	default:
+		// This case includes actual http transport errors. Go ahead and retry.
+	}
+	klog.Errorf("Unable to write event: '%v' (may retry after sleeping)", err)
+	return false
+
+}
+
+func createPatchBytes(event *v1beta1.Event) ([]byte, error) {
+
+	oldEvent := event.DeepCopy()
+	oldEvent.Series = nil
+
+	oldData, err := json.Marshal(oldEvent)
+	if err != nil {
+		return nil, err
+	}
+	newData, err := json.Marshal(event)
+	if err != nil {
+		return nil, err
+	}
+
+	return strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1beta1.Event{})
+
+}
+
+func getKey(event *v1beta1.Event) eventKey {
+	return eventKey{
+		action:              event.Action,
+		reason:              event.Reason,
+		reportingController: event.ReportingController,
+		reportingInstance:   event.ReportingInstance,
+		regarding:           event.Regarding,
+		related:             event.Related,
+	}
+}
+
+// StartEventWatcher starts sending events received from this EventBroadcaster to the given event handler function.
+// The return value can be ignored or used to stop recording, if desired.
+func (eventBroadcaster *eventBroadcasterImpl) StartEventWatcher(eventHandler func(event runtime.Object)) watch.Interface {
+	watcher := eventBroadcaster.Watch()
+	go func() {
+		defer utilruntime.HandleCrash()
+		for {
+			watchEvent, open := <-watcher.ResultChan()
+			if !open {
+				return
+			}
+			eventHandler(watchEvent.Object)
```

Comment:
I'm not really following the logic for this - there is significant different between e.g. add and delete of this watch event.
I think the interface for this eventHandler is a bad one - it should probably take watch.Event as a parameter.
Alternatively we may consider reusing the interface of ResourceHandlers from informer frameworks.

What is the usecase for it?

Reply:
> I'm not really following the logic for this - there is significant different between e.g. add and delete of this watch event.

Currently the eventrecorder only sends watch-events with add type.

> I think the interface for this eventHandler is a bad one - it should probably take watch.Event as a parameter.

I agree on this, we shouldn't loose the event

> Alternatively we may consider reusing the interface of ResourceHandlers from informer frameworks.
> 
> What is the usecase for it?

make event broadcaster able to watch Events created from the recorder, a couple of things that could be achieved with apimachinery watch:

- This would make us at some point able to shard to process generated events.

- I've seen clusters with dedicated event clusters + a very long TTL that want to clean up some events (e.g. events generated by some components). so we want to distinguish between creation and deletion at some point.